Get variables metadata for requested variables

JamesAnstey opened the pull request:
Output variables metadata in spreadsheet (issue #38) or json.

Usage example:
```
export_dreq_lists_json --all_opportunities v1.1 req.json -vm vars.json vars.csv --experiments piControl amip
```
produces
```
For data request version v1.1, number of requested variables found by experiment:
  amip : Core=139 ,High=835 ,Medium=203 ,Low=95, TOTAL=1272
  piControl : Core=139 ,High=924 ,Medium=307 ,Low=95, TOTAL=1465

Wrote requested variables to req.json
Wrote vars.json for 1587 variables, dreq version = v1.1
Wrote vars.csv for 1587 variables, dreq version = v1.1
```

Separate utility `get_variables_metadata.py` (currently in `scripts`) produces metadata file (json or csv), either for all variables (default) or filtered to some subset based on table name, variable name (out_name), or compound name. Usage example:

```
./get_variables_metadata.py -o vars.json vars.csv -t Amon SImon -v ua ta siconc sithick
```
produces
```
Retaining only these CMOR tables: Amon, SImon
Retaining only these CMOR variables: ua, ta, siconc, sithick
Wrote vars.json for 4 variables, dreq version = v1.1
Wrote vars.csv for 4 variables, dreq version = v1.1
```

JamesAnstey committed:
get_variables_metadata.py command-line interface and filtering by table,variable

JamesAnstey committed:
deleted 2 deprecated functions frmo dreq_query.py

JamesAnstey committed:
moved code that finds variable metadata into dreq_query.get_variables_metadata()

JamesAnstey committed:
write_cfsn_json.py to reorganize get_variables_metadata.py output json file by CF standard name

JamesAnstey committed:
added filtering by Compound Name to get_variables_metadata.py

JamesAnstey committed:
export_dreq_lists_json -vm option to produce variables metadata output file json or csv

matthew-mizielinski commented on data_request_api/command_line/export_dreq_lists_json.py:
```diff
@@ -28,6 +28,7 @@ def parse_args():
     parser.add_argument('--priority_cutoff', default='low', choices=dq.PRIORITY_LEVELS, help="discard variables that are requested at lower priority than this cutoff priority")
     parser.add_argument('output_file', help='file to write JSON output to')
     parser.add_argument('--version', action='store_true', help='Return version information and exit')
+    parser.add_argument('-vm', '--variables_metadata', nargs='+', type=str, help='output file(s) for metadata of requested variables (options: json, csv)')
```
The help here could do with a little clarification (I tried just putting in "csv" for this option and it did nothing :-)) 

I suggest changing the help to 
```
output files containing variable metadata of requested variables. Files with ".txt" or ".json" will be produced.
```

If you add the following code before defining this argument and change the type of this argument to `'txt_or_json_file'` Argparse will protect against the problem of file names slipping through the code;

```python
    def _var_metadata_check(arg):
        if not (arg.endswith('.txt') or arg.endswith('.json')):
            raise ValueError()

    parser.register('type', 'txt_or_json_file', _var_metadata_check)
```

JamesAnstey replied:
thanks @matthew-mizielinski - resolved by 15e59b6489fde7931efdc4bd1c7a19873943a6c6

matthew-mizielinski commented on data_request_api/stable/query/dreq_query.py:
I was slightly thrown when looking through this code to see variable names beginning with capital letters, e.g. `Frequency` as this usually means a class object.  Might be worth considering renaming variables to lower case to avoid this potential confusion point (you'd have to replace `Vars` with another name to avoid clashing with the `vars()` python function

JamesAnstey replied:
Assignments like this:
```
SpatialShape = base['Spatial Shape']
```
are objects of the `dreq_classes.dreq_table` class, so is the capitalization appropriate?

JamesAnstey replied:
oops, I misunderstood the term "class object", I thought it meant an instance of the class. Instances now renamed to user lowercase, and dreq_classes.py class names in CapWords following PEP8

matthew-mizielinski commented on data_request_api/stable/query/dreq_query.py:
```diff
@@ -711,200 +565,264 @@ def get_requested_variables(content, use_opps='all', priority_cutoff='Low', verb
+        if hasattr(var, 'cell_measures'):
+            # assert len(var.cell_measures) == 1
+            # link = var.cell_measures[0]
+            # cell_measures = CellMeasures.get_record(link).name
```
remove comments?

JamesAnstey replied:
done

matthew-mizielinski commented on scripts/get_variables_metadata.py:
```diff
@@ -16,320 +17,48 @@
+parser.add_argument('-t', '--cmor_tables', nargs='+', type=str,
+                    help='include only the specified CMOR tables (aka MIP tables, examples: "Amon", "Omon")')
+parser.add_argument('-v', '--cmor_variables', nargs='+', type=str,
+                    help='include only the specified CMOR variables (out_name, examples: "tas", "siconc")')
```
Consider whether any of these are mandatory (`dreq_version` and `outfile`?) as running the script with no arguments gives me an exception.

JamesAnstey replied:
ok, will adjust that

JamesAnstey replied:
resolved by ebdfe0e6420d8dfdc1ab116459eed11323b18bce

matthew-mizielinski commented on scripts/get_variables_metadata.py:
```diff
@@ -16,320 +17,48 @@
-###############################################################################
+default_dreq_version = 'v1.1'
 
+outfile_extensions = ['.json', '.csv']
```
It is possible to validate that these are included -- see comment above.

JamesAnstey replied:
resolved by 15e59b6489fde7931efdc4bd1c7a19873943a6c6

matthew-mizielinski reviewed (commented):
For consistency of interface for users could you consider moving the contents of the `get_variables_metadata.py` and `write_cfsn_json.py` scripts to `data_request_api.command_line`, moving all content inside a `main()` function and then updating the pyproject.toml file to point at these functions as "entry points"

JamesAnstey commented:
> For consistency of interface for users could you consider moving the contents of the `get_variables_metadata.py` and `write_cfsn_json.py` scripts to `data_request_api.command_line`, moving all content inside a `main()` function and then updating the pyproject.toml file to point at these functions as "entry points"

Yes, makes more sense in the cli part. Will revise as suggested and then move it there.

JamesAnstey committed:
fixed args.outfile validation

JamesAnstey committed:
made dreq_version a positional argument (no default), same as export_dreq_lists_json.py
made outfile a required named argument

JamesAnstey committed:
removed CapWords convention for dreq_table instances in dreq_query.get_variables_metadata()

JamesAnstey committed:
removed CapWords convention for other dreq_query functions, also cleaned up this module a little

JamesAnstey committed:
changed dreq_classes.py class names to CapWords

JamesAnstey committed:
moved write_cfsn_json.py to dev folder since not sure where it fits in yet

JamesAnstey committed:
moved get_variables_metadata.py to data_request_api.command_line, added pyproject.toml entry point

JamesAnstey commented:
> > For consistency of interface for users could you consider moving the contents of the `get_variables_metadata.py` and `write_cfsn_json.py` scripts to `data_request_api.command_line`, moving all content inside a `main()` function and then updating the pyproject.toml file to point at these functions as "entry points"
> 
> Yes, makes more sense in the cli part. Will revise as suggested and then move it there.

ok moved this to command_line dir and tested the pip install works ok e39c9473046290200990a5ce26454c20e25a6d12

JamesAnstey committed:
updated --variables_metadata argparse to be same as --outfile argparse in get_variables_metadata since they do the same thing

JamesAnstey committed:
merged main

JamesAnstey commented:
Thanks again @matthew-mizielinski for your review - I think all the points are addressed now. I merged in `main`, so hopefully this is ok to merge now. 

One thing: @sol1105 I noticed that the default config file has `consolidate : false`, which I needed to change to `true` for `get_variables_metadata` to run. This must be a bug in my code because it's supposed run for both cases. :)  So I'll look into that, but just wondering what the reason is to have `consolidate: false` by default?

JamesAnstey committed:
deleted old get_variables_metadata.py from scripts/ (replaced by command-line version)

JamesAnstey committed:
added coordinates (e.g. height2m) to dimensions metadata field produced by get_variables_metadata

JamesAnstey commented:
Added coordinates into the `dimensions` field of metadata output file. Mistakenly I left it out before, e.g. `3hrPt.vas` had
```
"dimensions": "longitude latitude time-point",
```
but it should be
```
"dimensions": "longitude latitude time-point height10m",
```
Fixed in 7fa3bdb56670394ccc22868666763558551eac45

JamesAnstey committed:
updated scripts/variable_info/all_var_info.json:
- coordinates included in 'dimensions'
- 'cmip6_cmor_table' renamed 'table' because in some cases the table name is not the same as CMIP6

sol1105 commented:
> One thing: @sol1105 I noticed that the default config file has `consolidate : false`, which I needed to change to `true` for `get_variables_metadata` to run. This must be a bug in my code because it's supposed run for both cases. :) So I'll look into that, but just wondering what the reason is to have `consolidate: false` by default?

The problem is that the consolidation is not yet working as it should and even resulted in critical errors, so I turned it off by default. The consolidation should 
- mimic Beth's steps when creating a release from the common airtable database , but iirc there were some limitations regarding the airtable exports from the public bases that caused problems here
- ensure consistency between versions (i.e. when fields get renamed between content releases)

I will try to get it working until our next software release, but 100% consistency is probably impossible to achieve, so it may also be easier to drop support of the old content versions (that should anyways be not eligible for data publication?!)
